Mathlib/Algebra/Lie/CartanExists.lean

```diff
@@ -0,0 +1,380 @@
+    have : ⁅x, z⁆ ∈ E := by rwa [← LieSubmodule.Quotient.mk_eq_zero']
+    -- From this we deduce that there exists an `n` such that `⁅x, _⁆ ^ n` vanishes on `⁅x, z⁆`.
+    -- On the other hand, our goal is to show `z = 0` in `Q`,
+    -- which is equivalent to showing that `⁅x, _⁆ ^ n` vanishes on `z`, for some `n`.
+    simp only [coe_bracket_of_module, LieSubmodule.mem_mk_iff', mem_coe_submodule, mem_engel_iff,
+      LieSubmodule.Quotient.mk'_apply, LieSubmodule.Quotient.mk_eq_zero', E, Q] at this ⊢
+    -- Hence we win.
+    obtain ⟨n, hn⟩ := this
+    use n+1
+    rwa [pow_succ]
+  -- Now we find a subset `s` of `K` of size `≥ r`
+  -- such that `constantCoeff ψ` takes non-zero values on all of `s`.
+  -- This turns out to be the subset that we alluded to earlier.
+  obtain ⟨s, hs, hsψ⟩ : ∃ s : Finset K, r ≤ s.card ∧ ∀ α ∈ s, (constantCoeff ψ).eval α ≠ 0 := by
+    classical
+    -- Let `t` denote the set of roots of `constantCoeff ψ`.
+    let t := (constantCoeff ψ).roots.toFinset
+    -- We show that `t` has cardinality at most `finrank K L - r`.
+    have ht : t.card ≤ finrank K L - r := by
+      refine (Multiset.toFinset_card_le _).trans ?_
+      refine (card_roots' _).trans ?_
+      rw [constantCoeff_apply]
+      -- Indeed, `constantCoeff ψ` has degree at most `finrank K Q = finrank K L - r`.
+      apply lieCharpoly_coeff_natDegree
+      suffices finrank K Q + r = finrank K L by rw [← this, zero_add, Nat.add_sub_cancel]
+      apply Submodule.finrank_quotient_add_finrank
+    -- Hence there exists a subset of size `≥ r` in the complement of `t`,
+    -- and `constantCoeff ψ` takes non-zero values on all of this subset.
+    obtain ⟨s, hs⟩ := exists_finset_le_card K _ hLK
+    use s \ t
+    refine ⟨?_, ?_⟩
+    · refine le_trans ?_ (Finset.le_card_sdiff _ _)
+      omega
+    · intro α hα
+      simp only [Finset.mem_sdiff, Multiset.mem_toFinset, mem_roots', IsRoot.def, not_and, t] at hα
+      exact hα.2 hψ
+  -- So finally we can continue our proof strategy by showing that `coeff χ i` vanishes on `s`.
+  apply eq_zero_of_natDegree_lt_card_of_eval_eq_zero' _ s _ ?hcard
+  case hcard =>
+    -- We need to show that `natDegree (coeff χ i) < s.card`
+    -- Which follows from our assumptions `i < r` and `r ≤ s.card`
+    -- and the fact that the degree of `coeff χ i` is less than or equal to `r - i`.
+    apply lt_of_le_of_lt (lieCharpoly_coeff_natDegree _ _ _ _ i (r - i) _)
+    · omega
+    · dsimp only [r] at hi ⊢
+      rw [Nat.add_sub_cancel' hi.le]
+  -- We need to show that for all `α ∈ s`, the polynomial `coeff χ i` evaluates to zero at `α`.
+  intro α hα
+  -- Once again, we are left with showing that `⁅y, _⁆` acts nilpotently on `E`.
+  rw [← coe_evalRingHom, ← coeff_map, lieCharpoly_map_eval,
+    (LinearMap.charpoly_eq_X_pow_iff _).mpr, coeff_X_pow, if_neg hi.ne]
+  -- To do so, it suffices to show that the Engel subalgebra of `y = a • u + x` is contained in `E`.
+  let y := α • u + x
```

Comment:
I was briefly confused by the fact that this shadows the existing `y`:
```suggestion
  -- To do so, it suffices to show that the Engel subalgebra of `v = a • u + x` is contained in `E`.
  let v := α • u + x
```
followed by a bunch of trivial fixes below ofc

Reply:
Thanks! Good suggestion.